File: parallel-consumer-proxy-clients/parallel-consumer-proxy-client-kotlin/src/main/kotlin/bz/stub/parallelconsumer/client/coroutines/Wire.kt
// Copyright (C) 2026 Antony Stubbs and contributors

package bz.stub.parallelconsumer.client.coroutines

import bz.stub.parallelconsumer.proxy.protocol.v1.Configure
import bz.stub.parallelconsumer.proxy.protocol.v1.Configured
import bz.stub.parallelconsumer.proxy.protocol.v1.DispatchRecord
import bz.stub.parallelconsumer.proxy.protocol.v1.ProduceRecord
import bz.stub.parallelconsumer.proxy.protocol.v1.Report
import bz.stub.parallelconsumer.proxy.protocol.v1.Token
import com.google.protobuf.ByteString
import java.time.Instant
import kotlin.time.Duration
import bz.stub.parallelconsumer.proxy.protocol.v1.ProcessingOrder as WireProcessingOrder

/**
 * The wire boundary: this client's own types in, protocol messages out, and back. Everything
 * protobuf-shaped funnels through here, so the session code above it reads in the library's terms -
 * and so the mapping is testable without a connection.
 *
 * **The token is echoed verbatim.** [toReport] sets the report's token to the very message object
 * the dispatch carried: no rebuild, no field read, no interpretation. Nothing in this client stores
 * anything per record - the token rides from the dispatch queue to the report on the executing
 * coroutine's stack, and a client that holds no per-record state cannot have a per-record state bug.
 */
internal object Wire {

    /**
     * The capability tokens this client actually implements, declared rather than left empty.
     *
     * An empty `capabilities` list does not mean "nothing" on the wire - the specification reads it
     * as the complete v1 baseline, which would claim heartbeats, manifest reconciliation,
     * worker-death reporting, terminal outcomes and the shutdown drain. Wave one performs exactly
     * one of those. The claim is not harmless: once the proxy grants `heartbeat`, a client that
     * never heartbeats has every in-flight record returned at lease expiry and its later reports
     * fenced as superseded, so nothing commits. Grow this set as each duty is genuinely
     * implemented, never ahead of it.
     */
    const val DISPATCH_CAPABILITY: String = "dispatch"

    /** The connect-time `Configure`: the options carried to the proxy, unmodified. */
    fun toConfigure(options: ClientOptions): Configure {
        val configure = Configure.newBuilder()
            .addAllTopics(options.topics)
            .addCapabilities(DISPATCH_CAPABILITY)
            .putAllKafkaProperties(options.kafkaProperties)
        options.maxConcurrency?.let { configure.maxConcurrency = it }
        options.ordering?.let { configure.ordering = toWireOrdering(it) }
        options.commitInterval?.let { configure.commitInterval = toWireDuration(it) }
        options.defaultMessageRetryDelay?.let { configure.defaultMessageRetryDelay = toWireDuration(it) }
        return configure.build()
    }

    /**
     * The handshake reply as this client's [Session] - the effective values, which is what the
     * client obeys. A `Configured` without the ceiling or the executor count is a protocol
     * violation, never an "unlimited": absence is meaningless for both.
     */
    fun toSession(configured: Configured): Session {
        if (!configured.hasMaxConcurrency() || !configured.hasExecutorCount()) {
            throw ProxyProtocolViolation(
                "the proxy's Configured omitted max_concurrency or executor_count; both are always set, " +
                    "and absence never means unlimited"
            )
        }
        return Session(
            executorCount = configured.executorCount,
            maxConcurrency = configured.maxConcurrency,
            capabilities = configured.capabilitiesList.toSet(),
        )
    }

    /**
     * One dispatched record as the user's function sees it. Absent wire fields become `null`:
     * Kafka's tombstone distinction for key and value, and "has not failed before" for the failure
     * history.
     */
    fun toInboundRecord(dispatch: DispatchRecord): InboundRecord {
        val record = dispatch.record
        return InboundRecord(
            topic = record.topic,
            partition = record.partition,
            offset = record.offset,
            key = if (record.hasKey()) record.key.toByteArray() else null,
            value = if (record.hasValue()) record.value.toByteArray() else null,
            attempt = if (dispatch.hasAttempt()) dispatch.attempt else 1,
            previousFailure = toPreviousFailure(dispatch),
        )
    }

    /** The outcome as a report, keyed by the dispatch's token echoed byte-identically. */
    fun toReport(token: Token, outcome: Outcome): Report {
        val report = Report.newBuilder().setToken(token)
        when (outcome) {
            is Outcome.Success -> report.success = toWireSuccess(outcome)
            is Outcome.Failure -> report.failure = toWireFailure(outcome)
        }
        return report.build()
    }

    private fun toPreviousFailure(dispatch: DispatchRecord): PreviousFailure? {
        if (!dispatch.hasLastFailureAt()) {
            return null
        }
        val at = Instant.ofEpochSecond(dispatch.lastFailureAt.seconds, dispatch.lastFailureAt.nanos.toLong())
        return PreviousFailure(at, if (dispatch.hasLastFailureReason()) dispatch.lastFailureReason else null)
    }

    private fun toWireSuccess(outcome: Outcome.Success): Report.Success {
        val success = Report.Success.newBuilder()
        outcome.produce.forEach { success.addProduce(toProduceRecord(it)) }
        return success.build()
    }

    private fun toWireFailure(outcome: Outcome.Failure): Report.Failure {
        val failure = Report.Failure.newBuilder()
        outcome.reason?.let { failure.reason = it }
        return failure.build()
    }

    private fun toProduceRecord(outbound: OutboundRecord): ProduceRecord {
        val produce = ProduceRecord.newBuilder().setTopic(outbound.topic)
        outbound.key?.let { produce.key = ByteString.copyFrom(it) }
        outbound.value?.let { produce.value = ByteString.copyFrom(it) }
        return produce.build()
    }

    private fun toWireOrdering(ordering: ProcessingOrder): WireProcessingOrder = when (ordering) {
        ProcessingOrder.UNORDERED -> WireProcessingOrder.PROCESSING_ORDER_UNORDERED
        ProcessingOrder.PARTITION -> WireProcessingOrder.PROCESSING_ORDER_PARTITION
        ProcessingOrder.KEY -> WireProcessingOrder.PROCESSING_ORDER_KEY
    }

    /** Built by hand rather than with protobuf-java-util, which is not on this module's classpath. */
    private fun toWireDuration(duration: Duration): com.google.protobuf.Duration {
        val wholeSeconds = duration.inWholeSeconds
        val nanos = duration.inWholeNanoseconds - wholeSeconds * NANOS_PER_SECOND
        return com.google.protobuf.Duration.newBuilder()
            .setSeconds(wholeSeconds)
            .setNanos(nanos.toInt())
            .build()
    }

    private const val NANOS_PER_SECOND = 1_000_000_000L
}

File: parallel-consumer-proxy-clients/parallel-consumer-proxy-client-python/src/parallel_consumer/_wire.py
# Copyright (C) 2026 Antony Stubbs and contributors

"""Translation between this library's surface and the frozen protocol's messages.

Kept in one module so the wire appears in exactly one place: nothing above it handles a protobuf
message, and nothing below it handles a :class:`~parallel_consumer.records.InboundRecord`.

The delivery token is the exception, and deliberately so. It travels as **opaque bytes** from
dispatch to report - out to the worker process and back - and is re-attached without either field
ever being read. That is what lets this client hold no request map, no completion registry and no
dedupe cache: it is stateless per record, and the fencing that needs the token's meaning is the
proxy's job.
"""

from __future__ import annotations

from datetime import timezone

from ._generated import proxy_pb2 as pb
from .options import ClientOptions, ProcessingOrder
from .outcomes import Outcome
from .records import InboundRecord

__all__ = ["configure_message", "inbound_record", "released_report", "report_message"]

_ORDERING = {
    ProcessingOrder.UNORDERED: pb.PROCESSING_ORDER_UNORDERED,
    ProcessingOrder.PARTITION: pb.PROCESSING_ORDER_PARTITION,
    ProcessingOrder.KEY: pb.PROCESSING_ORDER_KEY,
}

DISPATCH = "dispatch"
"""The one capability this wave implements. See :func:`configure_message`."""


def configure_message(options: ClientOptions, capabilities: tuple[str, ...]) -> pb.Configure:
    """Builds the connect-time ``Configure``.

    Capabilities are declared explicitly rather than left empty. Empty declares the complete v1
    baseline, which would claim duties this client does not yet perform; naming what it does
    implement lets the proxy hold it to exactly that.
    """
    configure = pb.Configure(kafka_properties=dict(options.kafka_properties),
                             capabilities=list(capabilities))
    if options.topics:
        configure.topics.extend(options.topics)
    if options.topic_pattern is not None:
        configure.topic_pattern = options.topic_pattern
    if options.max_concurrency is not None:
        configure.max_concurrency = options.max_concurrency
    if options.ordering is not None:
        configure.ordering = _ORDERING[options.ordering]
    if options.commit_interval is not None:
        configure.commit_interval.FromTimedelta(options.commit_interval)
    if options.default_message_retry_delay is not None:
        configure.default_message_retry_delay.FromTimedelta(options.default_message_retry_delay)
    return configure


def inbound_record(dispatched: pb.DispatchRecord) -> InboundRecord:
    """Turns one dispatched record into what the user's function sees.

    ``key`` and ``value`` keep Kafka's null/empty distinction: presence on the wire is what says
    the field was there at all, so a tombstone arrives as ``None`` and not as ``b""``.
    """
    record = dispatched.record
    return InboundRecord(
        topic=record.topic,
        partition=record.partition,
        offset=record.offset,
        key=record.key if record.HasField("key") else None,
        value=record.value if record.HasField("value") else None,
        attempt=dispatched.attempt,
        last_failure_at=(
            dispatched.last_failure_at.ToDatetime(tzinfo=timezone.utc)
            if dispatched.HasField("last_failure_at")
            else None
        ),
        last_failure_reason=(
            dispatched.last_failure_reason if dispatched.HasField("last_failure_reason") else None
        ),
    )


def report_message(token: bytes, outcome: Outcome) -> pb.Report:
    """Builds the ``Report`` for one resolved record, echoing the token verbatim."""
    report = pb.Report()
    report.token.ParseFromString(token)
    if outcome.succeeded:
        # SetInParent first: reading a oneof's submessage does not select the arm, and a bare
        # success (the common case) has nothing else to write that would select it.
        report.success.SetInParent()
        for outbound in outcome.produce:
            produce = report.success.produce.add()
            produce.topic = outbound.topic
            if outbound.key is not None:
                produce.key = outbound.key
            if outbound.value is not None:
                produce.value = outbound.value
    else:
        report.failure.SetInParent()
        if outcome.reason is not None:
            report.failure.reason = outcome.reason
    return report


def released_report(token: bytes) -> pb.Report:
    """Builds the ``Report`` for a record this client queued but never ran.

    Released is not a verdict: it returns the record to scheduling with its attempt count
    unchanged, which is how the client avoids inventing an outcome for work it did not do.
    """
    report = pb.Report()
    report.token.ParseFromString(token)
    report.released.SetInParent()
    return report
